Code/GraphMol/Descriptors/AtomicEnvironmentVector.cpp

```diff
@@ -0,0 +1,679 @@
+  index to encoding \param cutoff       Maximum distance within which 2 atoms
+  are considered to be neighbours
+
+  \return 2 dimensional array with 2 rows with each column corresponding to a
+  pair of atoms which are neighbours
+*/
+ArrayXd NeighborPairs(ArrayXXd coordinates, VectorXd species, double cutoff,
+                      int numAtoms) {
+  auto paddingMask = species.array() == -1;
+  int cols = 0;
+  MatrixXd upperTriag(2, numAtoms * (numAtoms - 1) / 2);
+  for (auto i = 0; i < numAtoms; i++) {
+    for (auto j = i + 1; j < numAtoms; j++) {
+      upperTriag.col(cols) << i, j;
+      cols++;
+    }
+  }
+
+  ArrayXd upperTriagFlattened(numAtoms * (numAtoms - 1));
+  int index = 0;
+  for (auto i = 0; i < upperTriag.rows(); i++) {
+    for (auto j = 0; j < upperTriag.cols(); j++) {
+      upperTriagFlattened(index) = upperTriag(i, j);
+      index++;
+    }
+  }
+
+  ArrayXXd pairCoordinates(numAtoms * (numAtoms - 1), 3);
+  pairCoordinates =
+      IndexSelect(pairCoordinates, coordinates, upperTriagFlattened, 0);
+
+  ArrayXXd distances(numAtoms * (numAtoms - 1) / 2, 3);
+  int numPairs = numAtoms * (numAtoms - 1) / 2;
+  for (auto i = 0; i < numAtoms * (numAtoms - 1) / 2; i++) {
+    distances.row(i) =
+        pairCoordinates.row(i) - pairCoordinates.row(i + numPairs);
+  }
+
+  auto distanceNorm = distances.matrix().rowwise().norm().array();
+
+  ArrayXXd pairPaddedMask(numAtoms * (numAtoms - 1), 1);
+  for (auto i = 0; i < upperTriagFlattened.size(); i++) {
+    pairPaddedMask(i, 0) = paddingMask(upperTriagFlattened(i), 0);
+  }
+
+  ArrayXd dist(distanceNorm.size());
+
+  for (auto i = 0; i < distanceNorm.size(); i++) {
+    if (pairPaddedMask(i, 0) == 1) {
+      dist(i) = INFINITY;
+    } else {
+      dist(i) = distanceNorm(i);
+    }
+  }
+
+  auto x = dist.array() <= cutoff;
+
+  std::vector<int> indices;
+  std::vector<std::pair<int, int>> atomIndex12;
+
+  for (auto i = 0; i < x.size(); i++) {
+    if (x(i) == 1) {
+      atomIndex12.push_back(
+          std::make_pair(upperTriag(0, i), upperTriag(1, i)));
+    }
+  }
+
+  ArrayXd atomIndex12Array(atomIndex12.size() * 2);
+  index = 0;
+  for (auto i : atomIndex12) {
+    atomIndex12Array(index) = i.first;
+    index++;
+  }
+  for (auto i : atomIndex12) {
+    atomIndex12Array(index) = i.second;
+    index++;
+  }
+
+  return atomIndex12Array;
+}
+
+ArrayXXd RadialTerms(double cutoff, ArrayXXd distances) {
+  auto fc = CosineCutoff(distances, cutoff);
+  ArrayXd ShfR(16);
+  ShfR << 0.9, 1.16875, 1.4375, 1.70625, 1.975, 2.24375, 2.5125, 2.78125, 3.05,
+      3.31875, 3.5875, 3.85625, 4.1250, 4.39375, 4.6625, 4.93125;
+  double EtaR = 16;
+  ArrayXXd ret(distances.rows(), 16);
+  for (auto i = 0; i < distances.rows(); i++) {
+    auto intermediate =
+        0.25 * ((ShfR - distances(i)).pow(2) * EtaR * -1).exp() * fc(i);
+    ret.row(i) = intermediate;
+  }
+  return ret;
+}
+
+ArrayXXd AngularTerms(double cutoff, ArrayXXd vectors12) {
```

Comment:
If there are any constraints on dimensions of `vectors12` they should be tested in a PRECONDITION